File: prototype/prototype.go
package prototype

type Cloneable interface {
	Clone() Cloneable
}

type PrototypeManager struct {
	manger map[string]Cloneable
}

func NewPrototypeManager() *PrototypeManager {
	return &PrototypeManager{
		manger: make(map[string]Cloneable),
	}
}

func (p PrototypeManager) Get(k string) Cloneable {
	return p.manger[k].Clone()
}

func (p PrototypeManager) Set(k string, v Cloneable) {
	p.manger[k] = v
}

File: prototype/cpp/prototype.cc
#include "prototype/cpp/prototype.h"
#include <string>

Cloneable::Cloneable(const std::string &name_):name(name_) {}

std::string Cloneable::get_name() {
    return this->name;
}


Cloneable1* Cloneable1::Clone() {
    return new Cloneable1(this->name);
}

void PrototypeManager::set(const std::string &key, Cloneable *value) {
    mp[key] = value;
}

Cloneable* PrototypeManager::get(const std::string &key) const {
    auto it = mp.find(key);
    return it != mp.end() ? it->second : nullptr;
}

PrototypeManager::~PrototypeManager() {
    for (auto i : mp) {
        if (i.second != nullptr) {
            delete i.second;
        }
    }
}
